**Student: Joel S. Mollel**

**Number: C00313599**

**Algorithm: Random Forest**

The purpose of this code is to implement a Random Forest classifier to predict drug types based on features such as Age, Sex, Blood Pressure, Cholesterol, and Na_to_K.

I used references to enhance my understanding of the decision tree and Random Forest algorithms, and these sources were instrumental in the implementation. The references I consulted are attached for your review. This project highlights my ability to preprocess data and apply machine learning techniques effectively to solve a classification problem.

To understand the algorithm well, after learning from the sources, I decided to implement it myself without using the provided code.

The Code is therefore my work and no changelog but steps taken to accomplish this work

Implementation of Random Forest Classifier for Classification of real life data

The project will use a dataset drug200.csv which is the labelled dataset that has various features. By using Age, sex, BP, cholestrol, Na-to-K ratio, the drug is prescribed to a patient. All the patients are suffering from same disease but the treatment dosage is different depending on the measurement of each feature for each patient.

Using the dataset, the model is trained with 70% of the dataset and tested with 30% of the dataset. The performance is evaluated based on accuracy.

**PART 1: Creating a basic Random Forest Classifier**

**STEPS**

*a) Importing important libraries and the dataset*

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

df = pd.read_csv('drug200.csv')

*b) Encoding categorical features since the model work with numbers*

In [2]:
df['Sex'] = df['Sex'].map({'F': 0, 'M': 1})  
df['BP'] = df['BP'].map({'HIGH': 0, 'LOW': 1, 'NORMAL': 2})  
df['Cholesterol'] = df['Cholesterol'].map({'HIGH': 0, 'NORMAL': 1}) 

*c) Feature and Target label preparation*

In [3]:
X = df[['Age', 'Sex', 'BP', 'Cholesterol', 'Na_to_K']]
y = df['Drug']

*d) Encoding target variable which is Drug*

In [4]:
encoder = LabelEncoder()
y_encoded = encoder.fit_transform(y)

*e) Splitting the data into training and testing sets, 70% and 30% respectively*

In [5]:
X_train, X_test, y_train, y_test = train_test_split(X, y_encoded, test_size=0.3, random_state=42)

*f) Creating and testing random Forest as Classifier*

In [6]:
model = RandomForestClassifier(n_estimators=100, random_state=42)
model.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

*g) Run predictions and evaluate model performance*

In [7]:
y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.4f}')

Accuracy: 1.0000


**Analysis of the Accuracy obtained(100%):** This accuracy is not reliable due to class imbalance issues and overfitting. It is therefore necessary that the two are addressed.

**PART 2: Introducing SMOTE and Cross Validation**

To enhance the accuracy of the model, overfitting and Class imbalance has to be addressed. SMOTE will address class imbalance while cross validation will address the overfitting.

*Step1: Loading the dataset and libraries*

In [8]:
import pandas as pd
from sklearn.model_selection import StratifiedKFold, cross_val_score, cross_val_predict
from sklearn.preprocessing import LabelEncoder
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from imblearn.over_sampling import SMOTE
from imblearn.pipeline import Pipeline
df = pd.read_csv('drug200.csv')

*Step2: Preparation and Encoding of categorical features and target*

In [9]:
df['Sex'] = df['Sex'].map({'F': 0, 'M': 1})  
df['BP'] = df['BP'].map({'HIGH': 0, 'LOW': 1, 'NORMAL': 2})  
df['Cholesterol'] = df['Cholesterol'].map({'HIGH': 0, 'NORMAL': 1}) 

# JM: Features and target variable definition
X = df[['Age', 'Sex', 'BP', 'Cholesterol', 'Na_to_K']]
y = df['Drug']

# JM: Label encoding for target variable
encoder = LabelEncoder()
y_encoded = encoder.fit_transform(y)

*Step 3: Initiation of the model and performing Cross validation of 5 folds while applying SMOTE*

In [10]:
# JM: Define the model
model = RandomForestClassifier(n_estimators=100, random_state=42)

# JM: Define SMOTE for handling class imbalance
smote = SMOTE(random_state=42)

# JM: Create a pipeline with SMOTE and the model
pipeline = Pipeline([('smote', smote), ('model', model)])

# JM: Define Stratified K-Fold cross-validation to preserve the class distribution
cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

# JM: Perform cross-validation to get predictions for each fold
y_pred = cross_val_predict(pipeline, X, y_encoded, cv=cv)

*Step 4: Evaluation of the model accuracy for each cross validation fold and the average accuracy*

In [11]:
# JM: Evaluate the accuracy for each fold
fold_accuracies = []
for fold, (train_idx, test_idx) in enumerate(cv.split(X, y_encoded)):
    fold_accuracy = accuracy_score(y_encoded[test_idx], y_pred[test_idx])
    fold_accuracies.append(fold_accuracy)
    print(f'Fold {fold + 1} Accuracy: {fold_accuracy:.4f}')

# JM: Calculate and print the average accuracy across all folds
avg_accuracy = sum(fold_accuracies) / len(fold_accuracies)
print(f'\nAverage Cross-Validation Accuracy: {avg_accuracy:.4f}')

# JM: Model training on the entire dataset for user input prediction
model.fit(X, y_encoded)

Fold 1 Accuracy: 1.0000
Fold 2 Accuracy: 1.0000
Fold 3 Accuracy: 1.0000
Fold 4 Accuracy: 1.0000
Fold 5 Accuracy: 0.9750

Average Cross-Validation Accuracy: 0.9950


RandomForestClassifier(random_state=42)

**Analysis of accuracy obtained:** The slight drop of accuracy from 100% to 99.5% is due to the fact that the 100% accuracy obtained earlier had overfitting and class imbalance issues. This accuracy(99.5%) is therefore comparatively more reliable than the earlier(100%)

**PART 3: Medical Drug Prescription System**

Using the model created, I am creating the drug prescription system. User inputs are processed and a drug is predicted based on the values provided for each feature prompted by the system.

In [12]:
# JM: User input for prediction
print('Dear Doctor,\n')
print("Please provide the following details of your patient to predict the drug:\n")
age = int(input("Age (e.g., 23): "))
sex = input("Sex (F/M): ")
bp = input("Blood Pressure (HIGH/LOW/NORMAL): ")
cholesterol = input("Cholesterol (HIGH/NORMAL): ")
na_to_k = float(input("Sodium to Potassium ratio (Na_to_K, e.g., 25.355): "))

# JM: Encoding the user input
sex_encoded = 0 if sex.upper() == 'F' else 1
bp_encoded = {'HIGH': 0, 'LOW': 1, 'NORMAL': 2}.get(bp.upper(), 2)
cholesterol_encoded = {'HIGH': 0, 'NORMAL': 1}.get(cholesterol.upper(), 1)

# JM: Creating a dataframe for the user input
user_input = pd.DataFrame([[age, sex_encoded, bp_encoded, cholesterol_encoded, na_to_k]], 
                          columns=['Age', 'Sex', 'BP', 'Cholesterol', 'Na_to_K'])

# JM: Predicting the drug for the user input
predicted_label = model.predict(user_input)
predicted_drug = encoder.inverse_transform(predicted_label)

# JM: Output the predicted drug
print("*" * 40)
print(f"The predicted drug for this patient based on given details is: {predicted_drug[0]}")

Dear Doctor,

Please provide the following details of your patient to predict the drug:



Age (e.g., 23):  23
Sex (F/M):  m
Blood Pressure (HIGH/LOW/NORMAL):  low
Cholesterol (HIGH/NORMAL):  normal
Sodium to Potassium ratio (Na_to_K, e.g., 25.355):  22


****************************************
The predicted drug for this patient based on given details is: drugY


**CONCLUSION**

The implementation of random Forest had a overall difficulty of 60% due to the earlier implementation of Decision Tree algorithm which followed similar procedure.

However, Random forest being an ensemble of many decision trees, it has shown more accuracy than one decision tree we implemented earlier. this lab has therefore proved the importance of ensemble methods in improving accuracy of the prediction.

After this lab, I am still continuing to explore deeper about decision trees and the application in the real life.

#### **REFERENCES**

Yahoo Finance. (2025). Tesla, Inc. (TSLA) Stock Price, News, Quote & History. Yahoo Finance, available at: https://finance.yahoo.com/quote/TSLA/?guccounter=1&guce_referrer=aHR0cHM6Ly93d3cuZ29vZ2xlLmNvbS8&guce_referrer_sig=AQAAAE0EIpCyfrPR0zR-jyMYOWNSqWOoYq3bdxTq2YmuXXmQO1JVPALSHNI0ZaHFjJygyLdh6qVqkW_gPDbzu_1NyGz1vrbq_ozMxjt4tGauDT1q4531B3vfwHJ-IrHQD15udeHq--64U9K1XuIlf8zYVeu2Oc11Czwp3XLfzGQQitSn (Accessed 12 Feb. 2025).

StatQuest with Josh Starmer (2016) Random Forests Part 1 - Building, Using, and Evaluating. Available at: https://www.youtube.com/watch?v=J4Wdy0Wc_xQ (Accessed: 12 February 2025).

Krish Naik (2021) Random Forest Algorithm In Detail | Machine Learning Algorithms. Available at: https://www.youtube.com/watch?v=nk2CQITm_eo (Accessed: 12 February 2025).

Scikit-learn (2023) 1.11. Ensemble methods — Random Forest - Scikit-learn Documentation. Available at: https://scikit-learn.org/stable/modules/ensemble.html#forests-of-randomized-trees (Accessed: 12 February 2025).

Marsland, S. (2015) Machine Learning: An Algorithmic Perspective. 2nd edn. Boca Raton, FL: CRC Press. Available at: https://www.routledge.com/Machine-Learning-An-Algorithmic-Perspective-Second-Edition/Marsland/p/book/9781466583283 (Accessed: 12 February 2025).

OpenAI (2025) ChatGPT, Available at: https://chat.openai.com/ (Accessed: 13 February 2025).
